### PR TITLE
feat: handle EIP-2612 and EIP-3009 in `SwapAndBridge.sol`

### PR DESCRIPTION
**Pull request by bmzig**
# Summary
We need to have a way to deposit into the spoke pool **and** perform a `swapAndBridge` using [EIP-2612](https://eips.ethereum.org/EIPS/eip-2612) and [EIP-3009](https://eips.ethereum.org/EIPS/eip-3009). 

This functionality is added with four new functions in the `UniversalSwapAndBridge` contract: `swapAndBridgeWithPermit`, `depositWithPermit`, `swapAndBridgeWithAuthorization`, and `depositWithAuthorization`, where `*WithPermit` indicates usage of an EIP-2612 compliant token, and `*WithAuthorization` indicates usage of an EIP-3009 compliant token.

# Details
EIP-2616 is used solely to give an approval to an address without calling `approve` first. This allows us to still use `transferFrom` when using any EIP-2612 token. 
On the other hand, EIP-3009 will simply transfer tokens to the designated receiver contract and not raise the contract's allowance, so all usage of these types of functions check whether a sufficient token amount was transferred to the contract before proceeding with the rest of the swap/deposit.

### CODE REVIEW

File: contracts/SwapAndBridge.sol [hunk 1]
```diff
@@ -291,4 +300,78 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
...
+            depositData,
+            IERC20(address(swapToken)), // Cast IERC20Permit to IERC20
+            acrossInputToken
+        );
+    }
+
+    /**
+     * @notice Swaps an EIP-3009 token on this chain via specified router before submitting Across deposit atomically.
+     * Caller can specify their slippage tolerance for the swap and Across deposit params.
+     // * @dev If swapToken does not implement `permit` to the specifications of EIP-2612, this function will fail.
```
Comment (bmzig) [inline contracts/SwapAndBridge.sol:343]: Need to edit this comment

File: contracts/SwapAndBridge.sol [hunk 2]
```diff
@@ -103,8 +104,13 @@ abstract contract SwapAndBridgeBase is Lockable, MultiCaller {
         // user on this contract, a malicious actor could call transferFrom to steal the user's tokens.
         if (!allowedSelectors[bytes4(routerCalldata)]) revert InvalidFunctionSelector();
 
-        // Pull tokens from caller into this contract.
-        _swapToken.safeTransferFrom(msg.sender, address(this), swapTokenAmount);
+        // Pull tokens from caller into this contract. This contract may already have tokens if, for example,
+        // a user sent the contract tokens and subsequently called `swapAndBridge`, so only pull funds if
+        // the contract needs to.
```
Comment (nicholaspai) [inline contracts/SwapAndBridge.sol:109]: This is a nice feature but curious why you're adding it to this PR, should it be in a separate PR?

Comment (bmzig) [reply contracts/SwapAndBridge.sol:109]: I added it here since `receiveWithAuthorization` should send tokens to the contract. So ideally, whenever `swapAndBridgeWithAuthorization` is called, `transferFrom` should never be called, and for the normal and permit swap and bridge, `transferFrom` should be called.

File: contracts/SwapAndBridge.sol (hunk 1 above)
Comment (bmzig) [reply contracts/SwapAndBridge.sol]: https://github.com/across-protocol/contracts/pull/751/commits/7bdf27a7ab2d78508e8e86ef519b5fdf63b99ff9

File: contracts/SwapAndBridge.sol (hunk 2 above)
Comment (nicholaspai) [reply contracts/SwapAndBridge.sol:110]: I see. Part of me thinks we should separate the token transfer logic from the swapping logic. Perhaps we should add a separate internal function called `_pullTokens` or something which has different behavior for `transferFrom` versus `permit` versus `receiveWithAuthorization `. What do you think?

Comment (bmzig) [reply contracts/SwapAndBridge.sol:110]: Yes, I think this will simplify things. I added a `_pullTokens` function in this commit: https://github.com/across-protocol/contracts/pull/751/commits/5f7ee7ba84f4b759ba9525a7d49dfd3455b434dc. I think the good thing about doing this is now all we really care about is whether the contract has sufficient balance to make the deposit/swap, and the way it gets the tokens is irrelevant.

Comment (nicholaspai) [reply contracts/SwapAndBridge.sol:110]: exactly

File: contracts/SwapAndBridge.sol [hunk 3]
```diff
@@ -103,8 +105,11 @@ abstract contract SwapAndBridgeBase is Lockable, MultiCaller {
         // user on this contract, a malicious actor could call transferFrom to steal the user's tokens.
         if (!allowedSelectors[bytes4(routerCalldata)]) revert InvalidFunctionSelector();
 
-        // Pull tokens from caller into this contract.
-        _swapToken.safeTransferFrom(msg.sender, address(this), swapTokenAmount);
+        // Pull tokens from caller into this contract. This contract may already have tokens if, for example,
+        // a user sent the contract tokens and subsequently called `swapAndBridge`, so only pull funds if
```
Comment (nicholaspai) [inline contracts/SwapAndBridge.sol:109]: I think this comment is a bit out of place now that we've added _pullTokens

File: contracts/SwapAndBridge.sol (hunk 2 above)
Comment (nicholaspai) [reply contracts/SwapAndBridge.sol:110]: so actually what I was suggesting is something more along the lines of removing this pulling logic from this function and leaving it the caller of `_swapAndBridge` to pull tokens before entering the function. `swapAndBridgeWithPermit`, `swapAndBridgeWithAuthorization` and `swapAndBridge` each have their own ways of pulling the user's funds. They can perform any checks they want to after pulling, and then they can call `_swapAndBridge`. What do you think of that? Am open to other ideas

File: contracts/SwapAndBridge.sol (hunk 3 above)
Comment (bmzig) [reply contracts/SwapAndBridge.sol]: Removed https://github.com/across-protocol/contracts/pull/751/commits/7d67ab56d693289fd6d71933b4a5084503139a07

File: contracts/SwapAndBridge.sol (hunk 2 above)
Comment (bmzig) [reply contracts/SwapAndBridge.sol]: Ah I see. So basically, leave this part to function in case we were to add more ways of depositing/authenticating. Makes sense. I'm just jumping around these different options. Under the impression that this contract will be modified more times in the future, I think it probably would be better this way. https://github.com/across-protocol/contracts/pull/751/commits/7d67ab56d693289fd6d71933b4a5084503139a07

File: contracts/SwapAndBridge.sol [hunk 4]
```diff
@@ -291,4 +310,178 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
...
+        uint256 validAfter,
+        uint256 validBefore,
+        bytes32 nonce,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external nonReentrant {
+        // While any contract can vacuously implement `transferWithAuthorization` (or just have a fallback),
+        // If tokens were not sent to this contract, the call to `transferFrom` in _swapAndBridge will revert.
+        swapToken.receiveWithAuthorization(
```
Comment (nicholaspai) [inline contracts/SwapAndBridge.sol:392]: Why can't you have `swapAndBridgeWithAuthorization`, `swapAndBridgeWithPermit`, and `swapAndBridge` all call:

```solidity
_pullTokens()
_swapAndBridge()
```
Where each of them defines `_pullTokens` differently..?

**Review by nicholaspai — COMMENTED** [review]
I don't think my message https://github.com/across-protocol/contracts/pull/751#discussion_r1847357534 came across clearly

File: contracts/SwapAndBridge.sol [hunk 5]
```diff
@@ -291,4 +310,178 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
...
+            validBefore,
+            nonce,
+            v,
+            r,
+            s
+        );
+        IERC20 _swapToken = IERC20(address(swapToken)); // Cast IERC20Auth to IERC20
+
+        uint256 tokenBalance = _swapToken.balanceOf(address(this));
+        if (tokenBalance < swapTokenAmount)
```
Comment (nicholaspai) [inline contracts/SwapAndBridge.sol:406]: Do you think we need this check? Won't `_swapAndBridge` fail if we have insufficient balance? Or do you think we want to just catch it earlier?

I guess i'm thinking, what's the difference between checking the balance here after calling `receiveWithAuthorization()` but not checking the balance after calling `safeTransferFrom` in the other two methods? Is it because in the latter we are more familiar with that function and knowing it **should just work**, whereas `receiveWithAuthorization` is a newer function that we're not used to?

Comment (bmzig) [reply contracts/SwapAndBridge.sol:406]: Yeah... That's a good point. I added this since I have the expectation that `receiveWithAuthorization` is just an uncommon function in most ERC20s, but whether a token implements this function can probably just be checked off-chain, so the check is just unnecessary.

File: contracts/SwapAndBridge.sol [hunk 6]
```diff
@@ -291,4 +327,171 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
...
+        uint256 minExpectedInputTokenAmount,
+        DepositData calldata depositData,
+        uint256 validAfter,
+        uint256 validBefore,
+        bytes32 nonce,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external nonReentrant {
+        // While any contract can vacuously implement `transferWithAuthorization` (or just have a fallback),
```
Comment (nicholaspai) [inline contracts/SwapAndBridge.sol:407]: is this comment line still relevant?

**Review by nicholaspai — APPROVED** [review]
Just one more Q on a comment!

File: contracts/SwapAndBridge.sol (hunk 6 above)
Comment (bmzig) [reply contracts/SwapAndBridge.sol:407]: It's still technically true. Maybe it's self explanatory, though. It's just to highlight that calling this function with a token that doesn't implement EIP-3009 will (probably) revert.

File: contracts/SwapAndBridge.sol [hunk 7]
```diff
@@ -291,4 +327,172 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
...
+        bytes calldata routerCalldata,
+        uint256 swapTokenAmount,
+        uint256 minExpectedInputTokenAmount,
+        DepositData calldata depositData,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external nonReentrant {
+        swapToken.permit(msg.sender, address(this), swapTokenAmount, deadline, v, r, s);
```
Comment (mrice32) [inline contracts/SwapAndBridge.sol:360]: What happens if the permit is frontrun?

Comment (bmzig) [reply contracts/SwapAndBridge.sol:360]: I think this call to permit will revert... So we should probably have a check beforehand which only calls permit if `allowance[msg.sender][address(this)]` is not enough to cover. Does that sound right?

File: contracts/SwapAndBridge.sol [hunk 8]
```diff
@@ -291,4 +328,174 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
...
+        uint256 swapTokenAmount,
+        uint256 minExpectedInputTokenAmount,
+        DepositData calldata depositData,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external nonReentrant {
+        IERC20 _swapToken = IERC20(address(swapToken)); // Cast IERC20Permit to IERC20.
+        if (_swapToken.allowance(msg.sender, address(this)) < swapTokenAmount)
```
Comment (nicholaspai) [inline contracts/SwapAndBridge.sol:362]: Why should we be checking for front-running with permit but not with transferFrom? I'd argue we'd want the same allowance() check in both swapAndBridge AND swapAndBridgeWithPermit.

But I favor having no checks at all and just letting the subsequent safeTransferFrom revert

Comment (bmzig) [reply contracts/SwapAndBridge.sol:362]: I thought the front-running was with respect to somebody calling `permit` on the token before this transaction is executed, so the approval would be there, but the call to permit would revert since the signature input in this function would have already been used.

Comment (nicholaspai) [reply contracts/SwapAndBridge.sol:362]: is this something that would happen in the normal course of transactions?

Comment (bmzig) [reply contracts/SwapAndBridge.sol]: It's only possible if people want to grief users. Updated here: https://github.com/across-protocol/contracts/pull/751/commits/ab96160a7580ed6a0bfb2aabe08b9b14ef16ded5

File: contracts/SwapAndBridge.sol [hunk 9]
```diff
@@ -291,4 +328,176 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
...
+        uint256 swapTokenAmount,
+        uint256 minExpectedInputTokenAmount,
+        DepositData calldata depositData,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external nonReentrant {
+        IERC20 _swapToken = IERC20(address(swapToken)); // Cast IERC20Permit to IERC20.
+        // For permit transactions, we wrap the call in a try/catch block so that the transaction will continue even if the call to
```
[the remark this answers is not in the document]
Comment (bmzig) [reply contracts/SwapAndBridge.sol:362]: https://github.com/across-protocol/contracts/pull/751/commits/7bc3b60d0b1b2bf475c41ddd5605597643687204

File: contracts/SwapAndBridge.sol [hunk 10]
```diff
@@ -79,11 +86,13 @@ abstract contract SwapAndBridgeBase is Lockable, MultiCaller {
      */
     constructor(
         V3SpokePoolInterface _spokePool,
+        WETH9Interface _wrappedNativeToken,
```
Comment (pxrl) [inline contracts/SwapAndBridge.sol:89]: @bmzig This is also able to be queried directly from the SpokePool contract. Thoughts on doing that within the constructor instead? It'd remove a constructor argument and would also eliminate the possibility for an (admittedly unlikely) misalignment on the address.

Comment (bmzig) [reply contracts/SwapAndBridge.sol:89]: I thought of doing that but `wrappedNativeTokens` isn't exposed by the V3SpokePoolInterface/SpokePoolInterface, so we would need to modify the interface for this to work (or just create a new interface). I figured we probably shouldn't do that.

File: contracts/SwapAndBridge.sol [hunk 11]
```diff
@@ -291,4 +328,178 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
             acrossInputToken
         );
     }
+
+    /**
+     * @notice Swaps an EIP-2612 token on this chain via specified router before submitting Across deposit atomically.
+     * Caller can specify their slippage tolerance for the swap and Across deposit params.
+     * @dev If swapToken does not implement `permit` to the specifications of EIP-2612, this function will fail.
```
Comment (pxrl) [inline contracts/SwapAndBridge.sol:335]: Is this comment still correct in light of the try/catch wrapper?

Comment (bmzig) [reply contracts/SwapAndBridge.sol:335]: It is. This comment is a bit more general, and just saying that you can't expect permit to work for a token which doesn't implement permit.

File: contracts/SwapAndBridge.sol [hunk 12]
```diff
@@ -291,4 +328,178 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
...
+            minExpectedInputTokenAmount,
+            depositData,
+            _swapToken,
+            acrossInputToken
+        );
+    }
+
+    /**
+     * @notice Deposits an EIP-2612 token Across input token into the Spoke Pool contract.
+     * @dev If `acrossInputToken` does not implement `permit` to the specifications of EIP-2612, this function will fail.
```
Comment (pxrl) [inline contracts/SwapAndBridge.sol:439]: Same q. RE the try/catch wrapper - is this comment still correct?

Comment (bmzig) [reply contracts/SwapAndBridge.sol:439]: You can think of this as: if a user uses a token which doesn't implement permit, then all else equal, this function will fail on `safeTransferFrom`, since `permit` does nothing to increase the approval. This is different from a token which does implement permit and has had their signature executed before the call to this function, since at least in that case, the allowance is expected to have changed.
